File: LeetCode/iter1/java/652.find-duplicate-subtrees.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class Solution {

    public static void main(String[] args) {
        testCase1();
        System.out.println();
    }

    private static void testCase1() {
        TreeNode t1 = new TreeNode(1);
        TreeNode t2 = new TreeNode(2);
        TreeNode t3 = new TreeNode(3);
        TreeNode t4 = new TreeNode(4);
        TreeNode t5 = new TreeNode(2);
        TreeNode t6 = new TreeNode(4);
        TreeNode t7 = new TreeNode(4);

        t5.left = t6;

        t3.left = t5;
        t3.right = t7;

        t2.left = t4;

        t1.left = t2;
        t1.right = t3;

        printLevelTree(t1);
        System.out.println();
        List<TreeNode> lst = findDuplicateSubtrees(t1);
        for (TreeNode root : lst) {
            printLevelTree(root);
            System.out.println();
        }
    }

    public static List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        Map<String, List<TreeNode>> serializeStr2Lst = new HashMap<>();
        serializeTreeNode(root, serializeStr2Lst);

        List<TreeNode> ret = new ArrayList<>();
        serializeStr2Lst.forEach((key, lst) -> {
            if (lst.size() > 1) {
                ret.add(lst.get(0));
            }
        });
        return ret;
    }

    private static String serializeTreeNode(TreeNode root, Map<String, List<TreeNode>> str2lst) {
        if (root == null) {
            return "#";
        }

        String serializeStr = root.val + serializeTreeNode(root.left, str2lst) + serializeTreeNode(root.right, str2lst);
        List<TreeNode> lst = str2lst.getOrDefault(serializeStr, new ArrayList<>());
        lst.add(root);
        str2lst.put(serializeStr, lst);
        return serializeStr;
    }

    private static void printLevelTree(TreeNode root) {
        Queue<Pair<TreeNode, Integer>> q = new LinkedList<>();
        q.add(new Pair<>(root, 1));
        int curLv = 1;
        while (!q.isEmpty()) {
            Pair<TreeNode, Integer> tmpPair = q.poll();
            if (tmpPair.getValue() != curLv) {
                curLv = tmpPair.getValue();
                System.out.println();
            }

            System.out.print(tmpPair.getKey().val + " ");
            if (tmpPair.getKey().left != null) {
                q.add(new Pair<>(tmpPair.getKey().left, tmpPair.getValue() + 1));
            }

            if (tmpPair.getKey().right != null) {
                q.add(new Pair<>(tmpPair.getKey().right, tmpPair.getValue() + 1));
            }
        }
        System.out.println();
    }

    static public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    static public class Pair<K, V> {
        private K key;
        private V value;

        Pair(K _key, V _value) {
            key = _key;
            value = _value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}

File: LeetCode/iter1/c++/652.find-duplicate-subtrees.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;


struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
		unordered_map<string, vector<TreeNode*>> map;
		serialize(root, map);
		vector<TreeNode*> ret;
		for (auto iter : map) {
			if (iter.second.size() > 1) {
				ret.push_back(iter.second[0]);
			}
		}
		return ret;
	}
private:
	string serialize(TreeNode* root, unordered_map<string, vector<TreeNode*>>& map) {
		if (root == nullptr) {
			return "#";
		}

		string serializeStr = to_string(root->val) + serialize(root->left, map) + serialize(root->right, map);
		if (map.count(serializeStr) == 1) {
			map[serializeStr].push_back(root);
		}
		else {
			vector<TreeNode*> v(1, root);
			map[serializeStr] = v;
		}

		return serializeStr;
	}
};

void testPrint(TreeNode* treeNode) {
	queue<pair<TreeNode*, int>> treeQ;
	treeQ.push(pair<TreeNode*, int>(treeNode, 1));
	int curLv = 1;

	while (!treeQ.empty()) {
		pair<TreeNode*, int>tmpPair = treeQ.front();
		treeQ.pop();

		if (curLv != tmpPair.second) {
			curLv = tmpPair.second;
			cout << endl;
		}

		cout << tmpPair.first->val << " ";
		if (tmpPair.first->left != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->left, tmpPair.second + 1));
		}
		if (tmpPair.first->right != NULL) {
			treeQ.push(pair<TreeNode*, int>(tmpPair.first->right, tmpPair.second + 1));
		}
	}
}

void testCase1() {
	TreeNode* t1 = new TreeNode(1);
	TreeNode* t2 = new TreeNode(2);
	TreeNode* t3 = new TreeNode(3);
	TreeNode* t4 = new TreeNode(4);
	TreeNode* t5 = new TreeNode(2);
	TreeNode* t6 = new TreeNode(4);
	TreeNode* t7 = new TreeNode(4);

	t5->left = t6;

	t3->left = t5;
	t3->right = t7;

	t2->left = t4;

	t1->left = t2;
	t1->right = t3;

	Solution s;
	vector<TreeNode*> v = s.findDuplicateSubtrees(t1);
	for (TreeNode* node : v) {
		testPrint(node);
		cout << endl;
		cout << endl;
	}
}

int main(void)
{
	testCase1();
	cout << endl;

	return 0;
}